src/mame/brother/lw30.cpp

```diff
@@ -0,0 +1,1105 @@
+	// write sector: (after sector_sync, 0xdd) sector_header (2+14), payload (416), sector_footer (11)
+
+	// from write_format, write_sector_data_header_data_footer
+	static constexpr int raw_sector_size = 8/*sector_prefix*/ + 2/*track_sync*/ + 2/*sector_sync*/ + 1/*0xdd*/ + 16/*sector_header*/ + 416/*payload*/ + 11/*sector_footer*/ + 42/*0xaa*/;
+	static constexpr int raw_track_size = 2/*0xaa*/ + 48/*0xaa*/ + sectors_per_track * raw_sector_size;
+
+	int identify(util::random_read& io, uint32_t form_factor, const std::vector<uint32_t>& variants) const override {
+		uint64_t size{};
+		io.length(size);
+		if(size == tracks_per_disk * sectors_per_track * sector_size)
+			return 50; // identified by size
+
+		return 0;
+	}
+
+	bool load(util::random_read& io, uint32_t form_factor, const std::vector<uint32_t>& variants, floppy_image* image) const override {
+		uint8_t trackdata[sectors_per_track * sector_size], rawdata[cells_per_rev / 8];
+		memset(rawdata, 0xaa, sizeof(rawdata));
+		for(int track = 0; track < tracks_per_disk; track++) {
+			size_t actual{};
+			io.read_at(track * sectors_per_track * sector_size, trackdata, sectors_per_track * sector_size, actual);
+			if(actual != sectors_per_track * sector_size)
+				return false;
+			size_t i = 0;
+			for(int x = 0; x < 2 + 48; x++)
+				rawdata[i++] = 0xaa;
+			auto interleave_offset = (track % 4) * 4;
+			for(size_t s = interleave_offset; s < interleave_offset + sectors_per_track; s++) {
+				auto sector = sector_interleave1[s % sectors_per_track] - 1;
+				// according to check_track_and_sector
+				for(const auto& d : sector_prefix) // 8 bytes
+					rawdata[i++] = d;
+				rawdata[i++] = sync_table[track] & 0xff;
+				rawdata[i++] = sync_table[track] >> 8;
+				rawdata[i++] = sync_table[sector] & 0xff;
+				rawdata[i++] = sync_table[sector] >> 8;
+				rawdata[i++] = 0xdd;
+				for(const auto& d : sector_header) // 16 bytes
+					rawdata[i++] = d;
+				auto payload = gcr_encode_and_checksum(trackdata + sector * sector_size); // 256 -> 416 bytes
+				for(const auto& d : payload)
+					rawdata[i++] = d;
+				for(const auto& d : sector_footer) // 11 bytes
+					rawdata[i++] = d;
+				for(int x = 0; x < 42; x++)
+					rawdata[i++] = 0xaa;
+			}
+			assert(i == raw_track_size);
+			assert(i <= cells_per_rev / 8);
+			generate_track_from_bitstream(track, 0, rawdata, cells_per_rev, image);
+		}
+
+		image->set_variant(floppy_image::SSDD);
+
+		return true;
+	}
+
+	const char* name() const override {
+		return "lw30";
+	}
+
+	const char* description() const override {
+		return "Brother LW-30 floppy disk image";
+	}
+
+	const char* extensions() const override {
+		return "img";
+	}
+
+	bool supports_save() const override {
+		// TODO
+		return false;
+	}
+};
+
+const lw30_format FLOPPY_LW30_FORMAT;
+
+class brother_beep_device : public device_t, public device_sound_interface
```

Comment:
Is this beeper used by other Brother machines?  If it is, it should be moved out to a separate file in mame/brother so it can be shared.

Reply:
This custom beeper device doesn't make much sense to me, almost all of it is a copy of sound/beep.*
Looking at the comments below (where it calls beeper_w), isn't the hardware just a 1bit dac?

Reply:
Nope, looking further, it's like a dac_2bit_ones_complement_device, where one bit is software controlled and the other is 4khz(can use clock_device for that)

Reply:
Unfortunately it's about 5 years since I reverse-engineered this hardware. As I remember, it uses a fixed 4 kHz beeper, but the Tetris game uses CPU-waitloop-controlled PWM to achieve different frequencies to play a melody. That didn't work with the original MAME beep device, so I modified it. Was something about restarting the oscillator or sth like this. Don't know about other Brother machines because I couldn't find any software on them to use it like the LW-30 Tetris.